### PR TITLE
Add task to check for invalid address

### PR DESCRIPTION
**Pull request by rioug**
#### What? Why?

As part of #12407, we added some migration that tighten the data validation at the database level, in particular for addresses. 
These migrations will fail deployment of the database contains invalid data.

This task check for invalid data addresses and check if they are in use, ie linked to another model. It will return :
- a list of invalid address in in use
- a list of invalid address not in use, ie that we can delete

Example output : 
```
$ rails ofn:data:check_invalid_address_used
Checking for invalid address
Checking if any of 66 invalid addresses are in use
"Customers []"
"Subscriptions []"
"EnterpriseGroup []"
"User []"
"Order []"
"Shipments []"
"Enterprises []"
"66 addresses can be deleted:"
[8965, 8669, 8670, 8752, 8753, 8872, 8873, 8914, 8915, 8958, 8959, 8961, 8962, 8964, 8967, 8968, 8991, 8992, 9015, 9016, 9048, 9049, 9128, 9129, 9088, 9089, 9158, 9159, 9193, 9194, 9418, 9419, 9327, 9328, 9527, 9528, 9563, 9564, 9570, 9571, 9572, 9573, 9586, 9587, 9591, 9592, 9594, 9595, 9620, 9621, 9624, 9625, 9658, 9659, 9716, 9717, 9719, 9720, 9726, 9727, 9730, 9731, 9734, 9735, 9762, 9763]
"0 addresses need to be fixed:"
[]
```

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Review by mkllnk — APPROVED** [review]
Very cool. We could add it to the migration to automatically delete.... too risky?

File: lib/tasks/check_invalid_address_used.rake [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      # - business_address
+      enterprises = Enterprise
+        .left_joins(:address, :business_address)
+        .where(
+          "address_id IN(?) OR business_address_id IN(?)", invalid_addresses, invalid_addresses
+        )
+      enterprise_used_address = enterprises.map do |e|
+        next e.address_id if check_correct_address_id(e.address_id, invalid_addresses)
+
+        e.business_address_id if check_correct_address_id(e.business_address_id, invalid_addresses)
```
Comment (dacook) [inline lib/tasks/check_invalid_address_used.rake:67]: Is it possible that an Enterprise has both invalid `address` _and_ `business_address`?

File: lib/tasks/check_invalid_address_used.rake [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    def check_bill_ship_address(klass, addresses)
+      objects = klass
+        .left_joins(:bill_address, :ship_address)
+        .where("bill_address_id in(?) OR ship_address_id in(?)", addresses, addresses)
+
+      objects.map do |o|
+        next o.ship_address_id if check_correct_address_id(o.ship_address_id, addresses)
+
+        o.bill_address_id if check_correct_address_id(o.bill_address_id, addresses)
```
Comment (dacook) [inline lib/tasks/check_invalid_address_used.rake:94]: As above, I think we could miss invalid addresses here.

**Review by dacook — CHANGES_REQUESTED** [review]
If my reading is right, I think we could potentially count some addresses as unused, when they are still used. Can you please check?

**Comment by rioug** [comment]
> Very cool. We could add it to the migration to automatically delete.... too risky?

Seems a bit risky, as an alternative I can printout the request/code to be run.

**Review by dacook — APPROVED** [review]
Cool, that should work 👍

**Pull request merged**